ecl/hql/hqlexpr.cpp

```diff
@@ -9143,6 +9338,122 @@ IHqlExpression * CHqlVirtualScope::lookupBaseSymbol(IHqlExpression * & definitio
+{
+    ForEachChild(i, this)
+    {
+        IHqlExpression * child = queryChild(i);
+        IHqlScope * base = child->queryScope();
+        if (base)
+        {
+            base->lookupNearestSymbol(searchName, lookupFlags|LSFfromderived, ctx, nearest);
+            if (nearest.bestMatchFound())
+                return nearest.querySymbol();
+        }
+    }
+
+    return nearest.querySymbol();
+}
+
+IHqlExpression * CHqlVirtualScope::lookupNearestSymbol(IIdAtom * searchName, unsigned lookupFlags, HqlLookupContext & ctx, NearestSymbol & nearest)
+{
+    //Are we just trying to find out what the definition in this scope is?
+    if (lookupFlags & LSFignoreBase)
+        return CHqlScope::lookupNearestSymbol(searchName, lookupFlags, ctx, nearest);
+
+    //The scope is complete=>this is a reference from outside the scope
+    if (complete)
+    {
+        //NOTE: If the members are virtual, then all that is significant is whether a match exists or not.
+        if (concrete && !(lookupFlags & LSFfromderived))
+            return concrete->lookupNearestSymbol(searchName, lookupFlags, ctx, nearest);
+
+        //The class is not concrete...
+        //1. It is based on a parameter, and not complete because the parameter hasn't been substituted.
+        //2. A reference from a derived module accessing the base definition.
+        //3. An illegal access to a member of an abstract class
+        NearestSymbol tempNearest(searchName);
+        IHqlExpression * match = CHqlScope::lookupNearestSymbol(searchName, lookupFlags, ctx, tempNearest);
+
+        if (!match)
+            return nearest.querySymbol();//NULL
+
+        if (!containsVirtual || (lookupFlags & LSFfromderived))
+        {
+            nearest.compare(tempNearest);
+            return nearest.querySymbol();
+        }
+
+        if (!isVirtualSymbol(match))
+        {
+            //Select from a parameter where the item is not defined in the parameter's scope => error
+            if (match->getOperator() != no_unboundselect)
+                nearest.compare(tempNearest);
+            return nearest.querySymbol();
+        }
+
+        if (canBeVirtual(match))
+        {
+            nearest.compare(tempNearest);
+            return nearest.querySymbol();
+        }
+
+        return nearest.querySymbol();
+    }
+
+    NearestSymbol scopeNearest(searchName);
+    IHqlExpression * scopeSymbol = CHqlScope::lookupNearestSymbol(searchName, lookupFlags, ctx, scopeNearest);
+
+    IHqlExpression * definitionModule = this;
+    NearestSymbol baseNearest(searchName);
+    IHqlExpression * baseSymbol = lookupNearestBaseSymbol(definitionModule, searchName, lookupFlags, ctx, baseNearest);
+
+    if (!scopeSymbol && !baseSymbol)
+        return nearest.querySymbol();//NULL
+
+    bool isScopeSymbolOK = false;
```

Comment:
This code doesn't make much sense.  Please ask if you don't know what a class represents.